draft-ietf-radext-radiusdtls-bis.md

```diff
@@ -624,6 +624,30 @@ This practice causes increased complexity in the client application and increase
+
+RADIUS/(D)TLS clients should use a single source (IP + port) when sending packets to a particular RADIUS/(D)TLS server. Doing so minimizes the number of (D)TLS session setups. It also ensures that information about the home server state is discovered only once.
+
+In practice, this means that RADIUS/(D)TLS clients with multiple internal RADIUS sources should use a local proxy that arbitrates all RADIUS traffic between the client and all servers. The proxy should accept traffic only from the authorized subsystems on the client machine and should proxy that traffic to known servers. Each authorized subsystem should include an attribute that uniquely identifies that subsystem to the proxy so that the proxy can apply origin-specific proxy rules and security policies. We suggest using NAS-Identifier for this purpose.
+
+The local proxy should be able to interact with multiple servers at the same time. There is no requirement that each server have its own unique proxy on the client, as that would be inefficient.
+
+The suggestion to use a local proxy means that there is only one process that discovers network and/or connectivity issues with a server. If each client subsystem communicated directly with a server, issues with that server would have to be discovered independently by each subsystem. The side effect would be increased delays in re-routing traffic, error reporting, and network instabilities.
+
+Each client subsystem can include a subsystem-specific NAS-Identifier in each request. The format of this attribute is implementation-specific. The proxy should verify that the request originated from the local system, ideally via a loopback address. The proxy MUST then rewrite any subsystem-specific NAS-Identifier to a NAS-Identifier that identifies the client as a whole, or remove the NAS-Identifier entirely and replace it with NAS-IP-Address or NAS-IPv6-Address.
+
+In traditional RADIUS, the cost to set up a new "session" between a client and server was minimal. The client subsystem could simply open a port, send a packet, wait for the response, and then close the port. With RADIUS/(D)TLS, the connection setup is significantly more expensive. In addition, there may be a requirement to use (D)TLS in order to communicate with a server, as other options may not be supported by that server. The knowledge of what protocol to use is best managed by a dedicated RADIUS subsystem rather than by each individual subsystem on the client.
+
+## Server Implementations
+
+RADIUS/(D)TLS servers should not use connected sockets to read (D)TLS packets from a client. This recommendation exists because a connected socket will accept packets only from one source IP address and port. This limitation would prevent the server from accepting packets from multiple clients on the same port.
```

Comment:
The use of unconnected sockets is only for UDP / DTLS.

Reply:
Where the src/dst ip/port is known in advance, connected UDP sockets work fine.  Where there is instead a "wildcard" socket for receiving packets from known clients but unknown source ports, connected UDP sockets have unavoidable race conditions in the network stack.

A RADIUS server would typically perform the following steps to get a connected socket:

1. open wildcard socket "*:1812" for UDP
2. accept packets for that socket
3. see new client source IP and source port
4. Open new UDP socket on the same source IP/port (SO_REUSEADDR / SO_REUSEPORT)
5. call connect() to connect to the client source IP and port
6. receive packets on the connected socket

For a TCP connection, steps (4) and (5) are handled by `accept()`, which is atomic.  It both opens a new socket, and ties that socket to a particular src/dst IP/port.

In contrast for UDP, steps (4) and (5) are done as separate system calls by the application.  There is no clear way to know what happens to packets during that window.

i.e. if there are multiple packets in the receive queue, does the `connect()` pull all packets out of the wildcard socket, and move them to the new socket?  Or do some packets go to the new connected socket, while other packets go to the original wildcard socket?

All of this is OS dependent.

If will _usually_ work, for the simple reason that the inter-packet spacing is larger than the time between system calls in steps (4) and (5).  But there is no _guarantee_ that it works.

If we do suggest that connected sockets are fine, then these issues should be explained, so that implementors are aware of them and can address them. 

This issue is address in more detail in a cloudflare article:  https://blog.cloudflare.com/everything-you-ever-wanted-to-know-about-udp-sockets-but-were-afraid-to-ask-part-1/

Reply:
I think here steps 4-5 are a non-issue. It may be in a general case (depending on the protocol, like quic), but since we are dealing explicitly with DTLS, its not.

Yes, between steps 3 and 5 any packet that arrives in that time frame _on the same 5-tuple_ will unpredictably be received by either the wildcard socekt or the connected socket. However in DTLS, there is no legal packat that can arrive _on the same 5-tuple_ between the initial client hello and an apropriate response from the server - which it will only send after connecting the socket. 

(sidenote: the initial client hello MUST fit into a single UDP packet; I stumbled over this when it didn't and OpenSSL would immediatley error out telling me so)

The worst thing that can happen is if duplicate UDP packets (client hello) arrive. If the connected socket gets it, I'm not shure how OpenSSL handles it, probably silently discards. If the wildcard socket gets it, it will try to connect another socket to the same 5-tuple, which will then fail.

Reply:
> However in DTLS, there is no legal packat that can arrive on the same 5-tuple between the initial client hello and an apropriate response from the server - which it will only send after connecting the socket.

Good point.  I agree then that DTLS connections SHOULD use connected sockets.

This issue should be described and explained in the spec.